File: example/BBoxLite/second_generation/C++/ConsoleApplication1/ConsoleApplication1/BBOXLITE_DEMO.cpp
#include "stdio.h"
#include "pch.h"

#include <iostream>
#include <string>
#include <array>
#include <msclr/marshal.h>
#include <msclr/marshal_cppstd.h>

using namespace System;
using namespace BBoxAPI;
using namespace msclr::interop;


int main()
{
	int dev_num = 0;
	BBoxAPI::retCode ^ ret;
	
	int STANDBY = 0;
	int TX = 1;
	int RX = 2;
	int SLEEP = 3;

	double TX_MIN_GAIN = 0;
	double TX_MAX_GAIN = 0;
	double RX_MIN_GAIN = 0;
	double RX_MAX_GAIN = 0;

	double TX_COM_MIN_GAIN = 0;
	double TX_COM_MAX_GAIN = 0;
	double RX_COM_MIN_GAIN = 0;
	double RX_COM_MAX_GAIN = 0;

	double TX_ELE_DR = 0;
	double RX_ELE_DR = 0;

	// GetDeviceStatus
	BBoxAPI::BBoxOneAPI ^instance = gcnew BBoxAPI::BBoxOneAPI();

	array<String ^>^ dev_info = instance->ScanningDevice((BBoxAPI::DEV_SCAN_MODE)0);
	dev_num = dev_info->Length;

	for (int i = 0; i < dev_num; i++)
	{
		array<String ^> ^ response_message = dev_info[i]->Split(',');	

		String ^ sn = response_message[0];
		std::string SN = msclr::interop::marshal_as<std::string>(sn);
		String ^ ip = response_message[1];
		std::string IP = msclr::interop::marshal_as<std::string>(ip);
		String ^ dev_type = response_message[2];
		std::string DEV_TYPE = msclr::interop::marshal_as<std::string>(dev_type);
		int devtype = std::stoi(DEV_TYPE);

		ret = instance->Init(sn, devtype, i);

		array<String ^> ^ freq_list = instance->getFrequencyList(sn);
		String ^ freq = freq_list[0];
		std::string FREQ = msclr::interop::marshal_as<std::string>(freq);
		double operating_freq = std::stod(FREQ);
		ret = instance->setOperatingFreq(operating_freq, sn);

		array<Double, 2> ^ DR = instance->getDR(sn);
		TX_MIN_GAIN = DR[0, 0];
		TX_MAX_GAIN = DR[0, 1];
		RX_MIN_GAIN = DR[1, 0];
		RX_MAX_GAIN = DR[1, 1];

		array<Double, 2> ^ COM_DR = instance->getCOMDR(sn);
		TX_COM_MIN_GAIN = COM_DR[0, 0];
		TX_COM_MAX_GAIN = COM_DR[0, 1];
		RX_COM_MIN_GAIN = COM_DR[1, 0];
		RX_COM_MAX_GAIN = COM_DR[1, 1];

		array<Double, 2> ^ ELE_DR = instance->getELEDR(sn);
		TX_ELE_DR = ELE_DR[0, 0];
		RX_ELE_DR = ELE_DR[0, 1];

		array<String ^> ^ AAkitList = instance->getAAKitList(sn);
		
		if (devtype == 7)
		{
			instance->selectAAKit("TMYTEK_28LITE_4x4_C2104L020-28", sn);
		}
		else if (devtype == 8)
		{
			instance->selectAAKit("TMYTEK_39LITE_4x4_A2104L004-39", sn);
		}

		Console::WriteLine("======================================================");
		Console::WriteLine("[DEMO1] Switch TX mode");
		system("pause");

		instance->SwitchTxRxMode(TX, sn);
		int mode = instance->getTxRxMode(sn);
		Console::WriteLine("[DEMO1] Mode : " + mode);

		Console::WriteLine("======================================================");
		Console::WriteLine("[DEMO2] Channel power control : Off channel 1 power");
		system("pause");

		int board = 1;
		int channel = 1;
		int sw = 1;
		instance->switchChannelPower(board, channel, sw, sn);
		Console::WriteLine("[DEMO2] Channel 1 power off");

		Console::WriteLine("======================================================");
		Console::WriteLine("[DEMO3] Channel Gain/Phase Control");
		Console::WriteLine("[DEMO3] Channel_1 gain  : 15 db");
		Console::WriteLine("[DEMO3] Channel_1 phase : 15 deg");
		Console::WriteLine("[DEMO3] Channel_2 gain  : 15 db");
		Console::WriteLine("[DEMO3] Channel_2 phase : 30 deg");
		Console::WriteLine("[DEMO3] Channel_3 gain  : 15 db");
		Console::WriteLine("[DEMO3] Channel_3 phase : 45 deg");
		Console::WriteLine("[DEMO3] Channel_4 gain  : 15 db");
		Console::WriteLine("[DEMO3] Channel_4 phase : 60 deg");
		system("pause");

		board = 1;

		channel = 1;
		double Target_db = 15;
		int Target_deg = 15;
		instance->setChannelGainPhase(board, channel, Target_db, Target_deg, sn);

		channel = 2;
		Target_db = 15;
		Target_deg = 30;
		instance->setChannelGainPhase(board, channel, Target_db, Target_deg, sn);

		channel = 3;
		Target_db = 15;
		Target_deg = 45;
		instance->setChannelGainPhase(board, channel, Target_db, Target_deg, sn);

		channel = 4;
		Target_db = 15;
		Target_deg = 60;
		instance->setChannelGainPhase(board, channel, Target_db, Target_deg, sn);

		Console::WriteLine("======================================================");
		Console::WriteLine("[DEMO4] BeamSteering Control");
		system("pause");
		Console::WriteLine("[DEMO4] Channel gain  : 12 db");
		Console::WriteLine("[DEMO4] Beam angle : 15 ");

		Target_db = 12;
		int Target_angle_x = 15;
		int Target_angle_y = 0;
		instance->setBeamXY(Target_db, Target_angle_x, Target_angle_y, sn);
	}

	Console::WriteLine("======================================================");
	Console::WriteLine("[DEMO] End");
	system("pause");
}

File: example/BBoxLite/second_generation/Python/BBOXLITE_DEMO.py
import sys
import clr
import time
import csv
import os
import numpy as np

dir_path = '..\\..\\..\\..\\'
os.chdir(os.path.abspath(dir_path))

path = '.\\BBoxAPI.dll'
clr.AddReference(os.path.abspath(path))


from BBoxAPI import *

if __name__ == '__main__':

	TX = 1
	RX = 2
 
	AAkitList = []

	instance = BBoxOneAPI()

	dev_info = instance.ScanningDevice(0)
	device_num = len(dev_info)

	if device_num > 0:
		for i in range(0, device_num, 1):

			response_message = dev_info[i].split(",")
			sn = response_message[0]
			ip = response_message[1]
			val = response_message[2].split("\x00")
			dev_type = int(val[0])

			print("[BBoxOne_Control_Interface][GetDeviceStatus] SN : %s" % (sn))
			print("[BBoxOne_Control_Interface][GetDeviceStatus] IP : %s" % (ip))
			print("[BBoxOne_Control_Interface][GetDeviceStatus] dev_type : %d" % (dev_type))
   
			instance.Init(sn, dev_type, i)
			print("[BBoxOne_Control_Interface][InitialDevice][%s]" % (sn))

			freq_list = instance.getFrequencyList(sn)
			if len(freq_list) == 0:
				print("[BBoxOne_Control_Interface][InitialDevice][%s] failed : return null" % (sn))
			else:
				for item in freq_list:
					print("[BBoxOne_Control_Interface][InitialDevice][%s] getFrequencyList: list %s" % (sn, item))

			ret = instance.setOperatingFreq(freq_list[0], sn)
			if ret != 0:
				print("[BBoxOne_Control_Interface][InitialDevice][%s] setOperatingFreq failed : error_code %d" % (sn, ret))

			DR = instance.getDR(sn)
			print("[BBoxOne_Control_Interface][InitialDevice][%s] " % (sn))
			TX_MIN_GAIN = DR[0, 0]
			print("[BBoxOne_Control_Interface][InitialDevice][%s] TX_MIN_GAIN : %f" % (sn, TX_MIN_GAIN))
			TX_MAX_GAIN = DR[0, 1]
			print("[BBoxOne_Control_Interface][InitialDevice][%s] TX_MAX_GAIN : %f" % (sn, TX_MAX_GAIN))
			RX_MIN_GAIN = DR[1, 0]
			print("[BBoxOne_Control_Interface][InitialDevice][%s] RX_MIN_GAIN : %f" % (sn, RX_MIN_GAIN))
			RX_MAX_GAIN = DR[1, 1]
			print("[BBoxOne_Control_Interface][InitialDevice][%s] RX_MAX_GAIN : %f" % (sn, RX_MAX_GAIN))

			AAkitList.append(instance.getAAKitList(sn))
			instance.selectAAKit(AAkitList[i][0], sn)

			print("======================================================")
			print("[DEMO1] Switch TX mode")
			os.system("pause")
			instance.SwitchTxRxMode(TX, sn)
			mode = instance.getTxRxMode(sn)
			print("[DEMO1] Mode : %d" % (mode))

			print("======================================================")
			print("[DEMO2] Channel power control : Off channel 1 power")
			os.system("pause")

			board = 1
			channel = 1
			sw = 1
			instance.switchChannelPower(board, channel, sw, sn)
			print("[DEMO2] Channel 1 power off")

			print("======================================================")
			print("[DEMO3] Channel Gain/Phase Control")
			print("[DEMO3] Channel_1 gain  : 15 db")
			print("[DEMO3] Channel_1 phase : 15 deg")
			print("[DEMO3] Channel_2 gain  : 15 db")
			print("[DEMO3] Channel_2 phase : 30 deg")
			print("[DEMO3] Channel_3 gain  : 15 db")
			print("[DEMO3] Channel_3 phase : 45 deg")
			print("[DEMO3] Channel_4 gain  : 15 db")
			print("[DEMO3] Channel_4 phase : 60 deg")
			os.system("pause")

			board = 1

			channel = 1
			Target_db = 15
			Target_deg = 15
			instance.setChannelGainPhase(board, channel, Target_db, Target_deg, sn)

			channel = 2
			Target_db = 15
			Target_deg = 30
			instance.setChannelGainPhase(board, channel, Target_db, Target_deg, sn)

			channel = 3
			Target_db = 15
			Target_deg = 45
			instance.setChannelGainPhase(board, channel, Target_db, Target_deg, sn)

			channel = 4
			Target_db = 15
			Target_deg = 60
			instance.setChannelGainPhase(board, channel, Target_db, Target_deg, sn)

			print("======================================================")
			print("[DEMO4] BeamSteering Control")
			os.system("pause")
			print("[DEMO4] Channel gain  : 12 db")
			print("[DEMO4] Beam angle : 15 ")

			Target_db = 12
			Target_angle_x = 15
			Target_angle_y = 0
			instance.setBeamXY(Target_db, Target_angle_x, Target_angle_y, sn)

	print("======================================================")
	print("[DEMO] End")
	os.system("pause")
